classes/i18n/LocaleMetadata.inc.php

```diff
@@ -0,0 +1,171 @@
+<?php
+declare(strict_types = 1);
+
+/**
+ * @defgroup i18n I18N
+ * Implements localization concerns such as locale files, time zones, and country lists.
+ */
+
+/**
+ * @file classes/i18n/LocaleMetadata.inc.php
+ *
+ * Copyright (c) 2014-2021 Simon Fraser University
+ * Copyright (c) 2000-2021 John Willinsky
+ * Distributed under the GNU GPL v3. For full terms see the file docs/COPYING.
+ *
+ * @class LocaleMetadata
+ * @ingroup i18n
+ *
+ * @brief Holds metadata about a system locale
+ */
+
+namespace PKP\i18n;
+
+use DomainException;
+use PKP\i18n\interfaces\LocaleInterface;
+use PKP\core\ExportableTrait;
+use PKP\core\PKPString;
+use PKP\facades\Locale;
+use Sokil\IsoCodes\Database\Languages\Language;
+
+class LocaleMetadata
+{
+    use ExportableTrait;
+
+    /** Locale identification */
+    public string $locale;
+
+    private ?object $_parsedLocale = null;
+
+    /**
+     * Constructor
+     */
+    public static function create(string $locale): self
+    {
+        $instance = new static();
+        $instance->locale = $locale;
+        return $instance;
+    }
+
+    /**
+     * Retrieves the display name
+     */
+    public function getDisplayName(?string $locale = null): string
+    {
+        $country = $this->getCountry($locale);
+        // Remove parenthesis from the language name and also convert the its character to uppercase (due to the IsoCodes texts)
+        return PKPString::regexp_replace('/\s*\([^)]*\)\s*/', '', PKPString::ucfirst($this->_getLanguage($locale)->getLocalName())) . ($country ? " (${country})" : '');
+    }
+
+    /**
+     * Retrieves the language name
+     */
+    public function getLanguage(?string $locale = null): string
+    {
+        return $this->_getLanguage($locale)->getLocalName();
+    }
+
+    /**
+     * Retrieves the country name
+     */
+    public function getCountry(?string $locale = null): ?string
+    {
+        return $this->_parse()->country ? Locale::getCountries($locale)->getByAlpha2($this->_parse()->country)->getLocalName() : null;
+    }
+
+    /**
+     * Retrieves the script name
+     */
+    public function getScript(?string $locale = null): ?string
+    {
+        $script = ucfirst($this->_parse()->script);
+        return $this->_parse()->script ? Locale::getScripts($locale)->getByAlpha4($script)->getLocalName() : null;
+    }
+
+    /**
+     * Whether the locale expects text on the right-to-left format
+     */
+    public function isRightToLeft(): bool
+    {
+        $locale = $this->_parse();
+        $language = strtolower($locale->language ?? '');
+        $script = strtolower($locale->script ?? '');
+        $rightToLeftLanguages = array_fill_keys(['ar', 'dv', 'fa', 'he', 'ku', 'nqo', 'prs', 'ps', 'sd', 'syr', 'ug', 'ur', 'yi'], true);
```

Comment:
Hmm, there were several benefits of the old `locales.xml` descriptor that we're losing here...
- Easy customization of the language display names,
- Easy determination of which languages were "complete enough" to even include in the installer UI,
- RTL vs. RTL status,
- Language-specific CSS file support

I'm not a fan of hard-coding lists like this, though admittedly this is one that is not likely to change often.

On the other hand, it's nice to have the completeness etc. automatically determined.

@NateWr, if I could ask a second opinion on this -- do you prefer descriptor or no descriptor?

Reply:
I think the RTL/CSS stuff is ok without a descriptor. That's not something that changes, really.

I do like the idea of allowing an installation to configure which locales it supports with a descriptor of some kind. We don't really have a pattern for this, but I've sometimes thought about having a directory in the `files_dir` where an install could place some files to override some of our registry files. (For example, if I as a host want a different set of default user groups, I could add a `files_dir/config/userGroups.xml` file with my own.) That's a use-case for keeping some of these registry files around where it is easy to do so, but I don't have a strong preference in this case.